### PR TITLE
tests/core/uc20-recovery: apply hack to get gopath in recover mode w/ external backend

### PR DESCRIPTION
**Pull request by anonymouse64**
With the external backend, we do not have the special snapd snap like we do on the google backend which copies gopath data from ubuntu-seed to the current ubuntu-data (which in recover mode is a tmpfs). Instead, we use a small hack, abusing the ~/.ssh/rc file which is run on the first ssh connection from spread, and bind mount the required files from /host/ubuntu-data to /home so that spread can run successfully.

We need to remove this ~/.ssh/rc file before going back to run mode however, as otherwise spread fails to re-connect with:
```
ssh: unexpected packet in response to channel open: <nil>
```
Unclear what makes that happen, as a normal ssh connection from i.e. a terminal works fine with this ~/.ssh/rc file in place.

### CODE REVIEW

**Comment by anonymouse64** [comment]
Best enjoyed with "Hide whitespace changes" to ignore the reformatting of this file

File: tests/core/uc20-recovery/task.yaml [hunk 1]
```diff
@@ -1,130 +1,168 @@
...
+            cat <<-EOF > /home/external/.ssh/rc
+    #!/bin/sh
+    # added by spread tests
+    if grep -q 'snapd_recovery_mode=recover' /proc/cmdline; then
+        # use as sudo for proper permissions, assumes that the external user is
+        # a sudo user, which it must be by definition of it being the login user
+        # that spread uses
+        # silence the output so that nothing is output to a ssh command,
+        # potentially confusing spread
+        sudo mkdir /home/gopath 2> /dev/null > /dev/null
```
Comment (bboozzoo) [inline tests/core/uc20-recovery/task.yaml:92]: According to sshd manpage, the script must not produce any output to stdout, but I'm not sure we should drop stderr here too. If those commands fail for any reason, it'll be hard to debug it.

File: tests/core/uc20-recovery/task.yaml [hunk 2]
```diff
@@ -1,130 +1,168 @@
...
+        # shellcheck disable=SC2016
+        retry -n 30 --wait 1 sh -c 'test "$(wc -l < /tmp/mock-shutdown.calls)" = "2"'
+        # an immediate reboot should have been scheduled
+        MATCH -- '-r \+0' < /tmp/mock-shutdown.calls
+
+        umount /usr/sbin/shutdown
+
+        # if we are using the external backend, remove the .ssh/rc hack we used
+        # to copy data around, it's not necessary going back to run mode, and it
+        # somehow interferes with spread re-connecting over ssh, causing an
```
Comment (bboozzoo) [inline tests/core/uc20-recovery/task.yaml:152]:
```suggestion
        # somehow interferes with spread re-connecting over ssh, causing
```

File: tests/core/uc20-recovery/task.yaml [hunk 3]
```diff
@@ -1,130 +1,168 @@
...
+
+        # restore shutdown so that spread can reboot the host
+        umount /usr/sbin/shutdown
+
+        # with the external backend, we do not have the special snapd snap with
+        # the first-boot run mode tweaks as created from $TESTLIB/prepare.sh's 
+        # repack_snapd_snap_with_deb_content_and_run_mode_firstboot_tweaks func
+        # so instead to get the spread gopath and other data needed to continue
+        # the test, we need to add a .ssh/rc script which copies all of the data
+        # from /host/ubuntu-data in recover mode to the tmpfs /home
```
Comment (bboozzoo) [inline tests/core/uc20-recovery/task.yaml:77]:
```suggestion
        # from /host/ubuntu-data in recover mode to the tmpfs /home, see
        # sshd(8) for details on the ~/.ssh/rc script
```

File: tests/core/uc20-recovery/task.yaml [hunk 4]
```diff
@@ -69,6 +69,25 @@ execute: |
         # restore shutdown so that spread can reboot the host
         umount /usr/sbin/shutdown
 
+        if [ "$SPREAD_BACKEND" = "external" ]; then
+            mkdir -p /home/external/.ssh
+            chown external:external /home/external/.ssh
+            touch /home/external/.ssh/rc
+            chown external:external /home/external/.ssh/rc
+            cat <<-EOF > /home/external/.ssh/rc
```
Comment (zyga) [inline tests/core/uc20-recovery/task.yaml:83]: Could you please add a reference to the `sshd(8)` manual page, specifically to the `sshrc` section.

**Pull request merged**